### PR TITLE
Add autoscale support

### PR DESCRIPTION
**Pull request by chefbob**
### Description

Add Chef resources for putting instances in and out of rotation

### Issues Resolved

Issue 348

### Check List

- [x] All tests pass. See <https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/TESTING.MD>
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] All commits have been signed for the Developer Certificate of Origin. See <https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/CONTRIBUTING.MD>

### CODE REVIEW

**Comment by chefbob** [comment]
Also includes WIP for get_parameters and get_parameters_by_path

File: test/fixtures/cookbooks/aws_test/recipes/ssm_parameter_store.rb [hunk 1]
```diff
@@ -71,12 +109,15 @@
   owner 'ec2-user'
   group 'ec2-user'
   mode '0755'
-  sensitive true
+  sensitive false
```
Comment (tas50) [inline test/fixtures/cookbooks/aws_test/recipes/ssm_parameter_store.rb]: False is the default so you can just remove this

File: libraries/matchers.rb [hunk 2]
```diff
@@ -198,7 +198,7 @@ def delete_aws_route53_record(resource_name)
 
   # System Manager Parameter Store
   def get_aws_ssm_parameter_store(resource_name)
-    ChefSpec::Matchers::ResourceMatcher.new(:aws_parameter_store, :create, resource_name)
+    ChefSpec::Matchers::ResourceMatcher.new(:aws_parameter_store, :get, resource_name)
```
Comment (tas50) [inline libraries/matchers.rb]: Just a future note. This whole file can go away. ChefSpec auto generates matchers now so there's no need to write them out by hand.

File: resources/autoscaling.rb [hunk 3]
```diff
@@ -0,0 +1,169 @@
+property :return_info, [String, Array]
+property :should_decrement_desired_capacity,	[true, false], default: true
+property :asg_name,              String
```
Comment (tas50) [inline resources/autoscaling.rb]: The spaces before the type should probably be consistent or just a single space. A single space would be my personal preference so we don't constantly have to update it when we add new properties.

File: resources/autoscaling.rb [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+# authentication
+property :aws_access_key,        String
+property :aws_secret_access_key, String
+property :aws_session_token,     String
+property :aws_assume_role_arn,   String
+property :aws_role_session_name, String
+property :region,                String, default: lazy { fallback_region }
+
+include AwsCookbook::Ec2 # needed for aws_region helper
```
Comment (tas50) [inline resources/autoscaling.rb]: This comment should actually read for the fallback_region helper

File: resources/autoscaling.rb [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+# allow use of the property names from the autoscaling cookbook
+alias_method :aws_access_key_id, :aws_access_key
+alias_method :aws_region, :region
+
+action :enter_standby do
```
Comment (tas50) [inline resources/autoscaling.rb]: The parts of these actions that actually make changes should be wrapped in a converge_by block. This allows for a few things

1) Friendly update messages in logs
2) Proper resource convergence for notifications
3) Support for why-run mode where chef will actually skip anything in the block

File: resources/ssm_parameter_store.rb [hunk 6]
```diff
@@ -6,6 +6,13 @@
 property :with_decryption,             [true, false], default: false
 property :allowed_pattern,             String
 property :return_key,                  String
+property :names,                       [String, Array], required: true
```
Comment (tas50) [inline resources/ssm_parameter_store.rb]: Was this previously required but not marked required?

**Comment by tas50** [comment]
Overall this looks like a really cool resource. There is a bit of rework that's necessary before it can be merged in. You'll want to poll the current state in each of the actions and then use converge_by to merge it only if necessary. This will make the resources actually idempotent. They shouldn't throw and error trying to change the state of a node if it's already in the requested state. That goes against how chef works and what users would expect. Let me know if you have any questions about that. Feel free to ping me under this same name on the community slack or track me down at ChefConf next week if you'll be there.

**Comment by chefbob** [comment]
Thanks Tim. I’ll be at ChefConf next week so I’ll look for you.

_______________
Bob DeRosa

On May 17, 2018, at 5:19 PM, Tim Smith <notifications@github.com<mailto:notifications@github.com>> wrote:


Overall this looks like a really cool resource. There is a bit of rework that's necessary before it can be merged in. You'll want to poll the current state in each of the actions and then use converge_by to merge it only if necessary. This will make the resources actually idempotent. They shouldn't throw and error trying to change the state of a node if it's already in the requested state. That goes against how chef works and what users would expect. Let me know if you have any questions about that. Feel free to ping me under this same name on the community slack or track me down at ChefConf next week if you'll be there.

File: test/fixtures/cookbooks/aws_test/recipes/ssm_parameter_store.rb (hunk 1 above)
Comment (chefbob) [reply test/fixtures/cookbooks/aws_test/recipes/ssm_parameter_store.rb]: changed it back to true.  I had changed it to false when I was testing.

File: resources/autoscaling.rb (hunk 3 above)
Comment (chefbob) [reply resources/autoscaling.rb]: done

File: resources/autoscaling.rb (hunk 4 above)
Comment (chefbob) [reply resources/autoscaling.rb]: done.

File: resources/autoscaling.rb (hunk 5 above)
Comment (chefbob) [reply resources/autoscaling.rb]: done

File: resources/ssm_parameter_store.rb (hunk 6 above)
Comment (chefbob) [reply resources/ssm_parameter_store.rb]: No.  names is new.  There used to be a name field but that was removed because it wasn't needed.

**Comment by chefbob** [comment]
@tas50 I updated the code based on your feedback.  I also managed to finish the get_parameters and get_parameters_by_path functions.  I'm currently getting a cookstyle error and I don't know how to solve it.  The error is 
resources/ssm_parameter_store.rb:49:17: C: Prefer to_s over string interpolation.
    secret_info["#{secret.name}"] = secret.value
                ^^^^^^^^^^^^^^^^
Here's the code
`action :get_parameters do
  request = {
    names: names,
    with_decryption: with_decryption,
  }
  resp = ssm_client.get_parameters(request)
  secret_info = {}
  resp.parameters.each do |secret|
    secret_info["#{secret.name}"] = secret.value
  end
  Chef::Log.debug "Get parameters #{names}"
  node.run_state[new_resource.return_keys] = secret_info
end
`

I tried using to_us and @ but am not able to get it working.

**Comment by chefbob** [comment]
fixed final cookstyle problem.  Thanks@bdwyertech.

**Pull request merged**